### ISSUE TITLE
Full screen resolution change in runtime

### ISSUE THREAD

**Issue by DimitriyPS** [godotengine/godot#2094]
Linux Mint 17.1 "Rebecca" Cinnamon, nVidia (additionally tested on Intel), Godot 1.1 (release)

I was trying to make user settings to change the screen resolution in the game for Godot. Just as is done in other PC games: http://savepic.org/7373870.htm

I tried to use the method: OS.set_video_mode(~), - but it's not working. Other tools I not found.

Why is this method not working?
Will it work in a future release?
Is there another way to switch the screen resolution?

**Comment by reduz** [comment]
this method will be removed soon, use set_window_size instead and related
methods

**Comment by DimitriyPS** [comment]
reduz, I am interested in full screen mode. I'm going to work with 3D. In full-screen mode set_window_size does not work.

**Comment by DimitriyPS** [comment]
I still can't understand. Task: using the full screen mode, switch the display resolution settings of the game developed in Godot. Godot has such an opportunity?

**Comment by reduz** [comment]
It won't hurt to look at the documentation.

Look, it's right next to set_window_size, the function I mentioned before:

https://github.com/okamstudio/godot/wiki/class_os

OS.set_window_fullscreen(bool)

**Comment by DimitriyPS** [comment]
Ok. I tried this method:

func _on_B1_pressed():
	if OS.is_video_mode_fullscreen(0)  !=  true:
		OS.set_window_fullscreen(true)
	else:
		OS.set_window_fullscreen(false)
	pass 

I have no problem that the program runs in full-screen mode. I am unable to change the resolution of this program. set_window_size - changes nothing. Maybe it's something I'm doing wrong.

**Comment by reduz** [comment]
Oh, custom resolution for fullscreen is deprecated, since most new
platforms don't support this.  Fullscren will always go full-res.
What you can do is to use stretch modes with viewport resize, so rendering
is made to smaller resolution and later stretched to fullscreen.
Technically, it's the same thing.

**Comment by DimitriyPS** [comment]
Perfectly. But I impossible to implement. I recently started to learn Godot. If it's not too much trouble, please make a small demo.

**Comment by reduz** [comment]
just set the desired game resolution in the project settings, and select
stretch mode 2D or viewport, all in the project settings menu

**Comment by DimitriyPS** [comment]
>>just set the desired game resolution in the project settings
It works. But it's not what necessary. This is not enough.

Games for PC needed switch permission. It is essential that:
1. The resolution of the PC monitor is not known in advance;
2. The user should be able to get the maximum quality on the monitor;
3. The user should be able to achieve higher productivity.

>>  What you can do is to use stretch modes with viewport resize, so rendering is made to smaller resolution and later stretched to fullscreen.

This is what need. Only need to be able to dynamically change the size of the renderer (which will then be stretched to fill the screen) using GDScript.

Is this possible?
You can implement this in version 1.2?

**Comment by okamstudio** [comment]
This is already implemented, check this tutorial:

https://github.com/okamstudio/godot/wiki/tutorial_multires

On Wed, Jun 17, 2015 at 5:03 PM, DimitriyPS <notifications@github.com>
wrote:

>  just set the desired game resolution in the project settings
> It works. But it's not what necessary. This is not enough.
>
>  Games for PC needed switch permission. It is essential that:
> 1. The resolution of the PC monitor is not known in advance;
> 2. The user should be able to get the maximum quality on the monitor;
> 3. The user should be able to achieve higher productivity.
>
>  What you can do is to use stretch modes with viewport resize, so
> rendering is made to smaller resolution and later stretched to fullscreen.
>
>  This is what need. Only need to be able to dynamically change the size
> of the renderer (which will then be stretched to fill the screen) using
> GDScript.
>
> Is this possible?
> You can implement this in version 1.2?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/okamstudio/godot/issues/2094#issuecomment-112930120>.
>



-- 

*OkamStudio*

**Comment by okamstudio** [comment]
if you want to do this from code using GDScript, check the SceneTree class,
which you can obtain from any node by calling get_tree(), but i don't think
it's needed, as the project settings options are enough.



On Wed, Jun 17, 2015 at 5:05 PM, Juan Linietsky <juan@okamstudio.com> wrote:

> This is already implemented, check this tutorial:
>
> https://github.com/okamstudio/godot/wiki/tutorial_multires
>
> On Wed, Jun 17, 2015 at 5:03 PM, DimitriyPS <notifications@github.com>
> wrote:
>
>>  just set the desired game resolution in the project settings
>> It works. But it's not what necessary. This is not enough.
>>
>>  Games for PC needed switch permission. It is essential that:
>> 1. The resolution of the PC monitor is not known in advance;
>> 2. The user should be able to get the maximum quality on the monitor;
>> 3. The user should be able to achieve higher productivity.
>>
>>  What you can do is to use stretch modes with viewport resize, so
>> rendering is made to smaller resolution and later stretched to fullscreen.
>>
>>  This is what need. Only need to be able to dynamically change the size
>> of the renderer (which will then be stretched to fill the screen) using
>> GDScript.
>>
>> Is this possible?
>> You can implement this in version 1.2?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/okamstudio/godot/issues/2094#issuecomment-112930120>.
>>
>
>
>
> --
>
> *OkamStudio*
>



-- 

*OkamStudio*

**Comment by DimitriyPS** [comment]
>> if you want to do this from code using GDScript, check the SceneTree class, which you can obtain from any node by calling get_tree()

okamstudio, тhank you. It seems to be working. To clarify... do I understand correctly that I need a method of "set_screen_stretch"? This will affect the 3D world?

extends Panel

var v1=Vector2(1366,768)
var v2=Vector2(800,600)

func _on_B1_pressed():
	get_tree().set_screen_stretch(true,true,v1)
	pass 
	
func _on_B2_pressed():
	get_tree().set_screen_stretch(true,true,v2)
	pass

Tell me, what do the first and second parameters (set_screen_stretch ( int mode, int aspect, - https://github.com/okamstudio/godot/wiki/class_scenetree#set_screen_stretch) of the method "set_screen_stretch"?

>>but i don't think it's needed, as the project settings options are enough.

It is a must! After all, the user will receive a finished game, and not its source. The user will not have the project settings. It is therefore necessary that these settings were in the game.

**Comment by Ogimle** [comment]
demo by link [working with multires in stretch mode](http://cgex.ru/download/file.php?id=874)

**Comment by kubecz3k** [comment]
To summarize this issue, it seems it's totally doable right now. Correct me if I'm wrong but the only thing that can be done here is to create demo or tutorial that illustrates how to give users possibility to change resolution of full-screened game in runtime.

**Comment by DimitriyPS** [comment]
Yes, it's really not a problem. If will be presented with a document with a official method of solving the problem is going to be great.

**Comment by vnen** [comment]
Window management demo allows you to resize on fullscreen, but the result is not good.

**Comment by reduz** [comment]
i fixed a few bugs regarding to keeping base resolution in ScreneTree, all should be good now

**Comment by nightblade9** [comment]
Sorry to revive an old thread. I spent quite a while researching this, and only now stumbled upon this example; the correct answer is buried in the comments above:

> just set the desired game resolution in the project settings, and select
stretch mode 2D or viewport, all in the project settings menu

Is it possible to create an example or put something in the docs about this? I ended up trying to get custom code to work, when all along, the core engine had what I needed.
